Get base news

san-e opened the pull request.

san-e committed:
base news

san-e committed:
add news() function to Base

biqqles commented on flint/missions.py:
```diff
@@ -52,6 +52,55 @@ def get_mbases() -> Dict[str, 'MBase']:
+				bases = [bases]
+
+			for base in bases:
+				if result.get(base):
+					result[base].append(NewsItem(**contents))
+				else:
+					result[base] = [NewsItem(**contents)]
```
Make `result = defaultdict(list)` and avoid needing to do this check

biqqles commented on flint/missions.py:
```diff
@@ -52,6 +52,55 @@ def get_mbases() -> Dict[str, 'MBase']:
+	for _, contents in news:
+		bases = contents.get("base")
+		if bases:
+			if type(bases) != list:
```
```suggestion
			if type(bases) is not list:
```

biqqles commented on flint/missions.py:
```diff
@@ -52,6 +52,55 @@ def get_mbases() -> Dict[str, 'MBase']:
+	base: List[str] = []
+
+	def lookup(self, id, markup = "html"):
+		_markup_formats = dict(html=dll.lookup_as_html, plain=dll.lookup_as_plain, rdl=dll.lookup)
```
Make this a class attribute.

Actually, we should find a way to refactor this dict out (probably put it in `interface.py`), since as you know it is defined in all these types of functions, but we can do that at a later date.

biqqles commented on flint/missions.py:
```diff
@@ -52,6 +52,55 @@ def get_mbases() -> Dict[str, 'MBase']:
+	def _category(self, markup = "html"):
+		return self.lookup(self.category, markup)
+
+	def _headline(self, markup = "html"):
```
I would guess that this (and/or the category) would be plaintext? Correct me if I am wrong, I haven't checked.

biqqles commented on flint/missions.py:
```diff
@@ -52,6 +52,55 @@ def get_mbases() -> Dict[str, 'MBase']:
+	rank: Optional[Tuple[str]] = ()
+	icon: Optional[str] = ""
+	logo: Optional[str] = ""
+	audio: Optional[bool] = False
```
These aren't optional if you provide a default value (unless `None` could be passed in from the ini parser, which never happens iirc).

biqqles commented on flint/missions.py:
```diff
@@ -52,6 +52,55 @@ def get_mbases() -> Dict[str, 'MBase']:
+	category: int
+	headline: int
+	text: int
+	rank: Optional[Tuple[str]] = ()
```
Nitpick: unless the tuple is specifically of length 1, the type annotation is `Tuple[str, ...]`. Not that intuitive, I know.

biqqles commented on flint/entities/universe.py:
```diff
@@ -122,6 +122,10 @@ def rumors(self, markup='html') -> Dict['Faction', Set[str]]:
         return {}
 
+    def news(self):
+        """A list of all news items being shown on this base"""
```
```suggestion
        """A list of all news items being shown on this base."""
```

biqqles reviewed (changes requested):
Thank you for another very nice feature! As well as the comments, minor stylistic details are that you should use spaces for indentation and single quotes for strings, to match the rest of the project.

biqqles commented:
Also sorry for the delay, I didn't see any notification from github about this.

Commit:
convert tab indent to spaces

Commit:
fix import

Commit:
make result defaultdict

Commit:
use single quotes

san-e committed:
.

Co-authored-by: biqqles <biqqles@protonmail.com>

Commit:
type() is not list

Commit:
Merge branch 'news' of https://github.com/san-e/flint into news

Commit:
remove Optional[]s

Commit:
fix Tuple type annotation

Commit:
even more single quotes

Commit:
replace lookup function with class attribute

biqqles committed:
Return empty list instead of None

biqqles committed:
Use name_ form, other formatting tweaks, more type hints

biqqles committed:
Use dll.lookup

Based on the fact no news item in Disco uses RDL
I am assuming that these fields are only
plain text. If this is wrong, I will revert
this commit.

biqqles commented:
Thanks very much! Excellent implementation. I made some minor stylistic changes. The methods you added I renamed to have the underscore at the end, since this matches elsewhere (e.g. `Base#system_` when names are taken by INI fields). I also used `dataclassy.Internal` to hide the `bases` attribute, since this is very long and makes the class neater to print.
